### PR TITLE
test(mock): implement a tiny mocking example

### LINKED ISSUE #14: State-of-the-art Rust mocking tools
## Aim

Provide a state-of-the-art for mocking tool with a tiny example.

## Assumption

- The mocking tool is important to avoid blocking in pipeline while other developers required a feature still in active development.
- It seems that other mocking tool are not under active development.
- We will only use Mockall

## Acceptance criteria

- [x] Fast and easy to use
- [x] It must be in active development
- [x] It must run safely with stable rust environment

## Risks

- None

## Dependencies

- [Mockall](https://crates.io/crates/mockall) crate

## References

- [mockall](https://crates.io/crates/mockall): it is a complete library that could mock structs, traits and functions. The documentation is detailed and comes with examples to quickly run it
- [A guide to mocking in Rust](https://blog.logrocket.com/guide-mocking-rust-mockall/): a little guide for various mocking tool in Rust

### PR DESCRIPTION
**Pull request by GuiBL4**
## Description

The state-of-the-art of mocking tool in Rust research could be sum up to the crate [Mockall](https://docs.rs/mockall/latest/mockall/index.html). It is the most maintained and up-to-date crate for mocking. It seems other crates are not well supported anymore.
This pull request (PR) implements a sample of greeter service mock.

This sample greeter service mock is mostly used as a mocking example.

## Related issues

This PR closes #14 

## How has this been tested?

- Not relevant

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

File: crates/greeter_mock/src/main.rs [hunk 1]
```diff
@@ -0,0 +1,45 @@
+//! Greeter mocking
+//!
+//! This crate implements a tiny greeter service mock using the crate 'mockall'
```
Comment (binadamu-isiyoonekana) [inline crates/greeter_mock/src/main.rs:3]: Good practice to hlink to the crate's GitHub repo (or crates.io hub), like https://github.com/asomers/mockall, for instance.

File: crates/greeter_mock/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,19 @@
+[package]
+name = "greeter_mock"
+description = "Sample of mocking using greeter"
```
Comment (binadamu-isiyoonekana) [inline crates/greeter_mock/Cargo.toml:3]: "Sample mocking for greeter service"

File: crates/greeter_mock/Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+documentation.workspace = true
+repository.workspace = true
+rust-version.workspace = true
+
+[[bin]]
+name = "greetermock"
+path = "src/main.rs"
+
+[dependencies]
+mockall = "0.11.3"
```
Comment (binadamu-isiyoonekana) [inline crates/greeter_mock/Cargo.toml:19]: mockall = { version = "0.11.3" }

File: crates/greeter_mock/src/main.rs [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+//!
+//! The scenario case is the following :
```
Comment (binadamu-isiyoonekana) [inline crates/greeter_mock/src/main.rs:5]: "The use case scenario is the following:"

File: crates/greeter_mock/src/main.rs [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+    let arguments: Vec<_> = env::args().collect();
+    if arguments.len() <= 1 {
+        println!("Who would you like to greet? Please give a name!");
+        println!("\nUsage: greetermock [name]");
+        exit(1);
+    }
+
+    // Call greeter service (first command-line argument is the name)
+    // You mock the trait here with a mutable initialization
+    let mut mock = MockGreeterService::new();
```
Comment (binadamu-isiyoonekana) [inline crates/greeter_mock/src/main.rs:29]: I would prefer 'service' (even if it is mocked) as a variable name.

**Review by binadamu-isiyoonekana — CHANGES_REQUESTED** [review]
Hello @GuiBL4. Thank you for your great work. Only tiny suggestions on the phrasing. Otherwise, the job is perfect. Bravo

File: Cargo.toml [hunk 6]
```diff
@@ -2,7 +2,8 @@
 
 members = [
   "crates/greeter_cli",
-  "crates/greeter_service"
+  "crates/greeter_service",
+  "crates/greeter_mock"
 ]
```
[the remark this answers is not in the document]
Comment (GuiBL4) [reply Cargo.toml:8]: You are right. After the state-of-the-art it appears there is only the Mockall crate which is up to date and maintained. This is the only mocking tool we will use.

**Review by binadamu-isiyoonekana — APPROVED** [review]
All is fine for me. Thanks @GuiBL4. Great work!

**Review by hmapedro — APPROVED** [review]

**Pull request merged**